Gitweb:     http://git.kernel.org/linus/be831297716036de5b24308447ecb69f1706a846
Commit:     be831297716036de5b24308447ecb69f1706a846
Parent:     9f800de38b05d84809e89f16671d636a140eede7
Author:     Joerg Roedel <joerg.roedel@amd.com>
AuthorDate: Mon Nov 23 12:50:00 2009 +0100
Committer:  Joerg Roedel <joerg.roedel@amd.com>
CommitDate: Mon Nov 23 12:54:17 2009 +0100

    x86/amd-iommu: attach devices to pre-allocated domains early
    
    For some devices the ACPI table may define unity map
    requirements which must me met when the IOMMU is enabled. So
    we need to attach devices to their domains as early as
    possible so that these mappings are in place when needed.
    This patch assigns the domains right after they are
    allocated. Otherwise this can result in I/O page faults
    before a driver binds to a device and BIOS is still using
    it.
    
    Cc: stable@kernel.org
    Signed-off-by: Joerg Roedel <joerg.roedel@amd.com>
---
 arch/x86/kernel/amd_iommu.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kernel/amd_iommu.c b/arch/x86/kernel/amd_iommu.c
index 093bd52..b74b212 100644
--- a/arch/x86/kernel/amd_iommu.c
+++ b/arch/x86/kernel/amd_iommu.c
@@ -2047,10 +2047,10 @@ static void prealloc_protection_domains(void)
 	struct pci_dev *dev = NULL;
 	struct dma_ops_domain *dma_dom;
 	struct amd_iommu *iommu;
-	u16 devid;
+	u16 devid, __devid;
 
 	while ((dev = pci_get_device(PCI_ANY_ID, PCI_ANY_ID, dev)) != NULL) {
-		devid = calc_devid(dev->bus->number, dev->devfn);
+		__devid = devid = calc_devid(dev->bus->number, dev->devfn);
 		if (devid > amd_iommu_last_bdf)
 			continue;
 		devid = amd_iommu_alias_table[devid];
@@ -2065,6 +2065,10 @@ static void prealloc_protection_domains(void)
 		init_unity_mappings_for_device(dma_dom, devid);
 		dma_dom->target_dev = devid;
 
+		attach_device(iommu, &dma_dom->domain, devid);
+		if (__devid != devid)
+			attach_device(iommu, &dma_dom->domain, __devid);
+
 		list_add_tail(&dma_dom->list, &iommu_pd_list);
 	}
 }
